_posts/2018-01-30-quicli.md

```diff
@@ -0,0 +1,294 @@
+	And while reading a file to a string this way may not be the most performant option,
+	it is very concise.
+	
+	Another aspect of conciseness:
+	Instead of using `.unwrap` to deal with errors,
+	quicli's `main!` allows you to use `?` to propagate errors,
+	and shows human-readable messages when your program exited with an error state.
+	Similarly, instead of
+	writing `return Err(SomeType::new("message"))`,
+	by using quicli you also get [failure]'s `bail!` and `ensure!`.
+	
+- well-structured and boilerplate-free
+
+	I just repeatedly wrote that this is a framework for _small_ code bases.
+	Nevertheless, it makes sense to structure your code well.
+	Actually, structuring your code in a clever way helps you keep the code size small.
+	
+	For example, this is how you parse CLI arguments (incl. type checks):
+	
+	```rust
+	/// Get first n lines of a file
+	#[derive(Debug, StructOpt)]
+	struct Cli {
+	    /// How many lines to get
+	    #[structopt(long = "count", short = "n", default_value = "3")]
+        count: usize,
+        /// The file to read
+	    file: String,
+    }
+    ```
+
+	See how you didn't write any code to convert whatever followed `--count` to an integer?
+	By the way:
+	Adding doc comments to your CLI arguments structure is not just for when you write the code,
+	structopt will also show them in a `--help`.
+
+- production-ready
+
+	In my experience,
+	the best way to ensure something ends up being used in production for years to come
+	is to call it a prototype.
+	So, I always want to get some basic stuff that will make using even the smallest tool less of a pain.
+
+	For example, CLI tools written with quicli (and structopt)
+	always have a `--help` message,
+	as well as `--version`,
+	and good error messages.
+	Similarly, you get logging for free,
+	and if you add two more lines,
+	you also get a `--verbose`/`-v` CLI argument to control the log level
+	(pass `-vvv` for "debug" level logging).
+	
+[failure]: https://docs.rs/failure
+
+## What I don't want quicli to be
+
+I want to use quicli to quickly build CLI apps
+– And I want to stop there!
+If you are no longer writing a small CLI tools
+(I'd say something like 500 lines of code is no longer small),
+your problems are most likely more complicated
+than what quicli gives you tools for.
+At that point,
+you should consider replacing your usage of (parts of) quicli
+with more sophisticated approaches.
+
+Let's contrast that with Garrett's [stdcli].
+It (re-)exports a lot of sophisticated tools!
+As Garrett writes:
+
+> stdcli's goal is basically to make creating a CLI in rust more like creating one in python from a user experience and ergonomics point of view.
+> This means:
+> batteries [are] included,
+> almost everything you need is already imported […]
+
+quicli also re-exports a bunch of stuff,
+and I agree that these re-exports make the experience of writing CLIs quickly much more smooth.
```

Comment:
I would replace much more (it's a valid but awkward construct) with either really or very. very sounds closer to what you want to imply.